### PR TITLE
ISPN-12868 Configure ACL cache

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.redhat.com/browse/ISPN-12868

### CODE REVIEW

File: core/src/main/java/org/infinispan/security/GlobalSecurityManager.java [hunk 1]
```diff
@@ -16,11 +18,12 @@
...
     */
-   Cache<?, ?> globalACLCache();
+   <K, V> Map<K, V> globalACLCache();
 
    /**
     * Flushes the ACL cache for this node
+    * @return
     */
-   void flushGlobalACLCache();
+   CompletionStage<Void> flushGlobalACLCache();
```
Comment (gustavonalle) [inline core/src/main/java/org/infinispan/security/GlobalSecurityManager.java:27]: This is public API, so we should deprecate rather than change it?

File: server/rest/src/main/java/org/infinispan/rest/resources/SecurityResource.java [hunk 2]
```diff
@@ -80,9 +82,18 @@ public Invocations getInvocations() {
...
             .handleWith(this::deny)
+            .invocation().methods(POST).path("/v2/security/cache").withAction("flush")
+            .permission(AuthorizationPermission.ADMIN).name("ACL CACHE FLUSH").auditContext(AuditContext.SERVER)
+            .handleWith(this::aclCacheFlush)
             .create();
    }
 
+   private CompletionStage<RestResponse> aclCacheFlush(RestRequest request) {
+      EmbeddedCacheManager cm = invocationHelper.getRestCacheManager().getInstance();
+      return SecurityActions.getGlobalComponentRegistry(cm).getComponent(GlobalSecurityManager.class).flushGlobalACLCache()
```
Comment (gustavonalle) [inline server/rest/src/main/java/org/infinispan/rest/resources/SecurityResource.java:93]: You can cache the ```GlobalSecurityManager``` instead of looking it up on every request I think

File: server/rest/src/main/java/org/infinispan/rest/resources/SecurityResource.java [hunk 3]
```diff
@@ -80,9 +82,18 @@ public Invocations getInvocations() {
             .invocation().methods(PUT).path("/v2/security/roles/{principal}").withAction("deny")
             .permission(AuthorizationPermission.ADMIN).name("ROLES DENY").auditContext(AuditContext.SERVER)
             .handleWith(this::deny)
+            .invocation().methods(POST).path("/v2/security/cache").withAction("flush")
+            .permission(AuthorizationPermission.ADMIN).name("ACL CACHE FLUSH").auditContext(AuditContext.SERVER)
+            .handleWith(this::aclCacheFlush)
```
Comment (gustavonalle) [inline server/rest/src/main/java/org/infinispan/rest/resources/SecurityResource.java:87]: Not strictly related to this PR, but well, let's not waste the context :)
Could add an indent or a line break between each invocation declaration to make it easier to read?

File: server/rest/src/main/java/org/infinispan/rest/resources/SecurityResource.java [hunk 4]
```diff
@@ -80,9 +82,18 @@ public Invocations getInvocations() {
...
             .create();
    }
 
+   private CompletionStage<RestResponse> aclCacheFlush(RestRequest request) {
+      EmbeddedCacheManager cm = invocationHelper.getRestCacheManager().getInstance();
+      return SecurityActions.getGlobalComponentRegistry(cm).getComponent(GlobalSecurityManager.class).flushGlobalACLCache()
+            .thenApply(v -> new NettyRestResponse.Builder().status(NO_CONTENT).build());
+   }
+
```
Comment (gustavonalle) [inline server/rest/src/main/java/org/infinispan/rest/resources/SecurityResource.java:96]: I couldn't find any documentation for the ```SecurityResource```, could you add it? It could be in another PR

**Review by gustavonalle — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/infinispan/security/GlobalSecurityManager.java (hunk 1 above)
Comment (tristantarrant) [reply core/src/main/java/org/infinispan/security/GlobalSecurityManager.java:27]: Not really: the only way to obtain a GlobalSecurityManager is through the component registry. I wouldn't worry about it

File: server/rest/src/main/java/org/infinispan/rest/resources/SecurityResource.java (hunk 2 above)
Comment (tristantarrant) [reply server/rest/src/main/java/org/infinispan/rest/resources/SecurityResource.java:93]: Will do

Comment (tristantarrant) [reply server/rest/src/main/java/org/infinispan/rest/resources/SecurityResource.java:93]: Rethinking about it: this is such a rare operation, that caching the GSM would be wasteful

**Comment by oraNod** [comment]
@tristantarrant FYI I pushed some suggested doc for the ACL cache.

**Comment by tristantarrant** [comment]
I've squashed the commits and added docs for the public security resource methods (I'd rather not document the methods which are of exclusive use to the console)

**Comment by ryanemerson** [comment]
Thanks @tristantarrant
